src/Microsoft.AspNet.OData.Shared/Builder/PropertyConfiguration.cs

```diff
@@ -157,6 +164,29 @@ public bool Unsortable
         /// </summary>
         public QueryConfiguration QueryConfiguration { get; set; }
 
+        /// <summary>
+        /// Adds TSubClass to the list of derived type constraints.
+        /// </summary>
+        internal PropertyConfiguration AddDerivedTypeConstraintImpl(Type t)
+        {
+            if (RelatedClrType == null)
+            {
+                throw Error.InvalidOperation("No CLR Type specified for this configuration.");
+            }
+            else if (!t.IsSubclassOf(RelatedClrType) && t != RelatedClrType)
+            {
+                throw Error.InvalidOperation("{0} is not a subclass of {1}. Cannot add this derived type constraint.", t.Name, RelatedClrType.Name);
+            }
+
+            if (DerivedTypeConstraints.Contains(t))
+            {
+                throw Error.InvalidOperation("Type {0} already added as derived type constraint", t.Name);
+            }
+
+            DerivedTypeConstraints.Add(t);
+            return this;
+        }
```

Comment:
It seems similar method repeated a lot, maybe consider to create a helper method